### ISSUE TITLE
Could not calculate PageRank during document imports

### ISSUE THREAD

**Issue by gabata** [arangodb/arangodb#5205]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [ ] 3.2
- [x] 3.3
- [ ] self-compiled devel branch

Mode:
- [ ] Cluster
- [x] Single-Server

Storage-Engine:
- [ ] mmfiles
- [x] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is a Pregel related issue
It seems that PageRank could not be calculated when a document import is currently in progress on the same collection where PageRank is intended to be calculated.

When I execute the following transaction via arangosh:

```javascript
db._executeTransaction({
  collections: { 
    read: ["nodes", "edges"],
    exclusive: "nodes",
    allowImplicit: true
  },
  action: function () {
    var internal = require("internal");
    var pregel = require('@arangodb/pregel');
    var params = {maxGSS: 5, threshold: 0.001};
    var handle = pregel.start('pagerank', 'graph', params);  
    while (!["done", "canceled", "in error"].includes(pregel.status(handle).state)) {
      internal.wait(0.5);
    }
    var status = pregel.status(handle);
    return status;
  }
});
```

I get the following response:

```json
{
  "state" : "done",
  "gss" : 1,
  "totalRuntime" : 72.33121609687805,
  "aggregators" : {
    "convergence" : -1
  },
  "sendCount" : 0,
  "receivedCount" : 0,
  "vertexCount" : 0,
  "edgeCount" : 558549
}
```

With the following log messages:
```
INFO {pregel} Running pregel with 0 vertices, 558549 edges
ERROR {threads} scheduler loop caught exception: internal error (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/Pregel/GraphStore.cpp:445). Please report this error to arangodb.com
ERROR {pregel} Pregel did not preallocate enough space for all edges. This hints at a bug with collection count()
ERROR {pregel} Storing vertex data failed
```

According to the documentation https://docs.arangodb.com/3.3/Manual/Transactions/LockingAndIsolation.html is it possible access collections in exclusive mode:
> In order to guard long-running or complex transactions against concurrent operations on the same data, the RocksDB engine allows to access collections in exclusive mode. Exclusive accesses will internally acquire a write-lock on the collections, so they are not executed in parallel with any other write operations. Read operations can still be carried out by other concurrent transactions.

Could you please give some support how it can be resolved?

Thanks,
Gabor

**Comment by gabata** [comment]
Hi all, may I ask if we could expect any deadline regarding that issue? If you have any suggestions for a workaround it would be also welcome. Thanks, Gabor

**Comment by graetzer** [comment]
I think there is no good workaround right now. It will currently be problematic to import documents while running pregel. I will try to implement a fix in the near term.

**Comment by gabata** [comment]
As a workaround we have implemented PageRank calculation via AQL:

```javascript
db._executeTransaction({
  collections: {
    write: ["nodes", "configuration"],
    read: ["nodes", "edges", "configuration"]
  },
  action: function () {
    var internal = require('internal');
    var arango = require('@arangodb');
    var runFlag = 'configuration/PAGE_RANK_CALCULATION_RUNNING_FLAG';
    var nodes = 'nodes';
    var edges = 'edges';
    var iterations = 4;

    var initialPageRankStep = [
      'LET initialRank = 1.0 / LENGTH(@@nodes)',
      'FOR i IN @@nodes',
      'LET outCount = FIRST(FOR v IN 1..1 OUTBOUND i @@edges COLLECT AGGREGATE c = COUNT(v) RETURN c)',
      'UPDATE i._key WITH { pageRank: initialRank, outCount: outCount } IN @@nodes',
      'OPTIONS { ignoreErrors: true }'
    ].join(' ');

    var pageRankIterationStep = [
      'LET d = 0.85',
      'LET n = LENGTH(@@nodes)',
      'FOR i IN @@nodes',
      'LET p = (1.0 - d) / n + d * FIRST(FOR v IN 1..1 INBOUND i @@edges COLLECT AGGREGATE sumRank = SUM(v.rank / v.outCount) RETURN sumRank)',
      'UPDATE i._key WITH { pageRank: p } IN @@nodes',
      'OPTIONS { ignoreErrors: true }'
    ].join(' ');

    var status = arango.db._exists(runFlag) && arango.db._document(runFlag);
    if (!!status && !status.value) {
      arango.db._update(status._id, { value: true });
      internal.wait(60);
      console.log('[aql-pagerank] Set initial pageRank values for ' + nodes);
      arango.db._query(initialPageRankStep, { '@edges': edges, '@nodes': nodes });

      for (var i = 1; i <= iterations; i++) {
        console.log('[aql-pagerank] Calculate pageRank, iteration: ' + i);
        arango.db._query(pageRankIterationStep, { '@edges': edges, '@nodes': nodes });
      }
    } else {
      console.log('[aql-pagerank] Calculation has already been started.');
    }
    arango.db._update(status._id, { value: false });
    console.log('[aql-pagerank] Done.');
  }
});
```

**Comment by maxkernbach** [comment]
Hi @gabata,

 ArangoDB 3.4 RC2, which includes a fix for this issue (see PR https://github.com/arangodb/arangodb/pull/6356), has been released and can be found [here](https://www.arangodb.com/download-technical-preview/).

Could you please check if the bug is fixed in 3.4 RC2?

**Comment by gabata** [comment]
Hi @maxkernbach, thanks for the fix. Unfortunately, I don't have chance to try and verify that at the moment, please feel free to close this ticket.

Now we are using a workaround: stop importing when Pregel is started. It also turned out, that there was a correlation between https://github.com/arangodb/arangodb/issues/5205 and this issue.

**Comment by maxkernbach** [comment]
@gabata 

Thanks for your reply. Since no stable version including this fix is available just yet I will keep this issue open until the 3.4 GA release.

It seems that you linked a wrong issue in your latest comment (#5205 is the current issue).

**Comment by gabata** [comment]
@maxkernbach sorry the issue I originally wanted to mention is https://github.com/arangodb/arangodb/pull/6039

**Comment by maxkernbach** [comment]
Closing - fixed in 3.4.0-RC.2
Please feel free to comment in case the problem persists and we will reopen this issue.
